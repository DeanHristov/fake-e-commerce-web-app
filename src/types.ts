export enum PRODUCT_CATEGORIES {
  ELECTRONIC = 'ELECTRONIC',
}

export type TCurrency = 'USD' | 'BGN' | 'EUR';

export interface IComment {
  productId: string;
  _id: string;
  name: string;
  email: string;
  body: string;
  date: string;
}

// Object schema can be seen here: https://dummyjson.com/docs/products
export interface IProduct {
  id: number;
  title: string;
  description: string;
  price: number;
  discountPercentage: number;
  rating: number;
  stock: number;
  brand: string;
  category: string;
  thumbnail: string;
  images?: string[];
  comments?: IComment[];
}

export interface IUser {
  name: string;
  email: string;
  password: string;
  isAdmin: boolean;
}

export interface ICartProduct {
  id: number;
  price: number;
  quantity: number;
  discountPercentage?: number;
  // TODO Fix next line!
  discountedPrice?: number;
  title: string;
  thumbnail: string;
  total: number;
  inStock: number;
}

export interface ICart {
  id: number;
  userId: number;
  total: number;
  discountedTotal: number;
  totalProducts: number;
  totalQuantity: number;
  currency?: TCurrency;
  products: Record<number, ICartProduct>; // Map<number, ICartProduct>;
}

//@see:https://nextjs.org/docs/app/api-reference/functions/generate-metadata#generatemetadata-function
export interface IMetadataPageTitle {
  title: string;
}

export interface IProductsResponse {
  products: IProduct[];
  total: number;
  skip: number;
  limit: number;
}

export interface ICartsResponse {
  carts: ICart[];
  total: number;
  skip: number;
  limit: number;
}
