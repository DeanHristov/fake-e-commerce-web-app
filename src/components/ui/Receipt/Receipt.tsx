import { FC } from 'react';
import Divider from '@/components/ui/Divider';
import { ICartProduct, TCurrency } from '@/types';
import { Utils } from '@/utils/Utils';
import format from 'date-fns/format';

export interface IReceiptProps {
  total: number;
  currency: TCurrency;
  products: ICartProduct[];
}

const Receipt: FC<IReceiptProps> = ({ total, currency, products }) => {
  return (
    <div className="cepeipt-container px-4 space-y-6 my-6 pb-6 border border-gray-300 rounded-md">
      <div className="py-2">
        <h3 className="text-center text-slate-900 text-3xl">Fake App</h3>
        <p className="text-center text-sm text-gray-400">
          {format(new Date(), 'dd MMMM, yyyy')}
        </p>
      </div>
      <div className="">
        {products.map(({ title, total, quantity }, index) => (
          <p key={`receipt-item-${index}`} className="receipt-item py-2">
            <span className="text-slate-900 text-sm bg-white px-1 z-10">
              {`(${quantity}x) ${title}`}
            </span>
            <span className="text-slate-900 text-sm bg-white px-1 z-10">
              {Utils.parseAmountByCurrency(total, currency)}
            </span>
          </p>
        ))}
      </div>
      <Divider gradient />
      <p className="receipt-item py-1">
        <span className="text-slate-900 text-sm font-bold bg-white px-1 z-10">
          Total
        </span>
        <span className="text-slate-900 text-sm font-bold bg-white px-1 z-10">
          {Utils.parseAmountByCurrency(total, currency)}
        </span>
      </p>
      <h3 className="text-center text-slate-900 text-2xl">Thank you!</h3>
    </div>
  );
};

export default Receipt;
