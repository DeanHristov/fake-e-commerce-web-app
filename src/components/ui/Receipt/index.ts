import Receipt, { IReceiptProps } from './Receipt';

export type { IReceiptProps };
export default Receipt;
