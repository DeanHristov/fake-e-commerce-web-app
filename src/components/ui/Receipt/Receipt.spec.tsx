import { render } from '@testing-library/react';
import Receipt from '@/components/ui/Receipt';
import { ICartProduct } from '@/types';

describe('UI/Component <Receipt {...} />', () => {
  it('Should be able to render the render the default state', () => {
    const { container } = render(
      <Receipt total={20} currency={'EUR'} products={[] as ICartProduct[]} />,
    );

    expect(container).toMatchSnapshot();
  });
});
