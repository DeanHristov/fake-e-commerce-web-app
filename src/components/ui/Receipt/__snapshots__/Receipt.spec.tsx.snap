// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI/Component <Receipt {...} /> Should be able to render the render the default state 1`] = `
<div>
  <div
    class="cepeipt-container px-4 space-y-6 my-6 pb-6 border border-gray-300 rounded-md"
  >
    <div
      class="py-2"
    >
      <h3
        class="text-center text-slate-900 text-3xl"
      >
        Fake App
      </h3>
      <p
        class="text-center text-sm text-gray-400"
      >
        17 December, 2023
      </p>
    </div>
    <div
      class=""
    />
    <div
      class="w-full h-px m-auto bg-gradient-to-r from-transparent via-gray-400 to-transparent"
    />
    <p
      class="receipt-item py-1"
    >
      <span
        class="text-slate-900 text-sm font-bold bg-white px-1 z-10"
      >
        Total
      </span>
      <span
        class="text-slate-900 text-sm font-bold bg-white px-1 z-10"
      >
        â‚¬20.00
      </span>
    </p>
    <h3
      class="text-center text-slate-900 text-2xl"
    >
      Thank you!
    </h3>
  </div>
</div>
`;
